File: backend/src/services/task.service.ts
/**
 * Task Service
 * Business logic for task management
 */

import { prisma } from '../models';
import { Task, TaskStatus } from '@prisma/client';
import { TaskCreateInput, TaskUpdateInput, VALID_STATUS_TRANSITIONS } from '../models/task.model';
import { NotFoundError, ConflictError } from '../utils/errors';
import logger from '../utils/logger';

export class TaskService {
  /**
   * Create a new task
   */
  async createTask(data: TaskCreateInput): Promise<Task> {
    const task = await prisma.task.create({
      data: {
        title: data.title,
        description: data.description,
        status: data.status || 'Pending',
      },
    });

    logger.info('Task created:', { taskId: task.id, title: task.title });

    // Log activity
    await this.logActivity(task, 'task.created');

    return task;
  }

  /**
   * Get task by ID
   */
  async getTask(id: string): Promise<Task> {
    const task = await prisma.task.findUnique({
      where: { id },
    });

    if (!task) {
      throw new NotFoundError('Task');
    }

    return task;
  }

  /**
   * Get all tasks with optional filtering
   */
  async getTasks(filters?: {
    status?: TaskStatus;
    page?: number;
    pageSize?: number;
  }): Promise<{ tasks: Task[]; total: number; page: number; pageSize: number }> {
    const { status, page = 1, pageSize = 20 } = filters || {};

    const where = status ? { status } : {};

    const [tasks, total] = await Promise.all([
      prisma.task.findMany({
        where,
        orderBy: { createdAt: 'desc' },
        skip: (page - 1) * pageSize,
        take: pageSize,
      }),
      prisma.task.count({ where }),
    ]);

    return { tasks, total, page, pageSize };
  }

  /**
   * Update task
   */
  async updateTask(id: string, data: TaskUpdateInput): Promise<Task> {
    // Verify task exists
    await this.getTask(id);

    const updateData: Record<string, unknown> = {
      ...data,
    };

    // Handle completedAt based on status
    if (data.status === 'Done') {
      updateData.completedAt = new Date();
    } else if (data.status !== undefined) {
      updateData.completedAt = null;
    }

    const task = await prisma.task.update({
      where: { id },
      data: updateData,
    });

    logger.info('Task updated:', { taskId: task.id });
    await this.logActivity(task, 'task.updated');

    return task;
  }

  /**
   * Update task status with transition validation
   */
  async updateTaskStatus(id: string, newStatus: TaskStatus): Promise<Task> {
    const task = await this.getTask(id);

    // Validate status transition
    const validTransitions = VALID_STATUS_TRANSITIONS[task.status];

    if (!validTransitions.includes(newStatus)) {
      throw new ConflictError(
        `Cannot transition from ${task.status} to ${newStatus}. Valid transitions: ${validTransitions.join(', ') || 'none'}`,
        'INVALID_TRANSITION'
      );
    }

    const updateData: Record<string, unknown> = {
      status: newStatus,
    };

    if (newStatus === 'Done') {
      updateData.completedAt = new Date();
    } else {
      updateData.completedAt = null;
    }

    const updatedTask = await prisma.task.update({
      where: { id },
      data: updateData,
    });

    logger.info('Task status updated:', {
      taskId: task.id,
      fromStatus: task.status,
      toStatus: newStatus,
    });

    await this.logActivity(updatedTask, 'task.updated', {
      fromStatus: task.status,
      toStatus: newStatus,
    });

    return updatedTask;
  }

  /**
   * Delete task
   */
  async deleteTask(id: string): Promise<void> {
    // Verify task exists
    await this.getTask(id);

    await prisma.task.delete({
      where: { id },
    });

    logger.info('Task deleted:', { taskId: id });

    // Log activity (best effort, task is already deleted)
    try {
      await prisma.activityLog.create({
        data: {
          type: 'task.deleted',
          description: `Task ${id} deleted`,
          metadata: { taskId: id },
        },
      });
    } catch (error) {
      logger.warn('Failed to log task deletion:', { taskId: id, error });
    }
  }

  /**
   * Log activity for task operations
   */
  private async logActivity(
    task: Task,
    type: string,
    additionalMetadata?: Record<string, unknown>
  ): Promise<void> {
    try {
      await prisma.activityLog.create({
        data: {
          type,
          description: `Task '${task.title}' ${this.getActivityDescription(type)}`,
          metadata: additionalMetadata ? JSON.parse(JSON.stringify(additionalMetadata)) : undefined,
        },
      });
    } catch (error) {
      logger.warn('Failed to log activity:', { taskId: task.id, type, error });
    }
  }

  /**
   * Get activity description based on type
   */
  private getActivityDescription(type: string): string {
    const descriptions: Record<string, string> = {
      'task.created': 'created',
      'task.updated': 'updated',
      'task.deleted': 'deleted',
    };
    return descriptions[type] || 'modified';
  }
}

export const taskService = new TaskService();
export default taskService;

File: backend-python/app/services/task_service.py
from datetime import datetime
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession
from app.models.task import Task
from app.models.activity_log import ActivityLog

VALID_STATUS_TRANSITIONS: dict[str, list[str]] = {
    "Pending": ["In Progress", "Done", "Cancelled"],
    "In Progress": ["Done", "Pending", "Cancelled"],
    "Done": [],
    "Cancelled": ["Pending"],
    "PENDING": ["RUNNING", "COMPLETED", "FAILED", "SKIPPED"],
    "RUNNING": ["COMPLETED", "FAILED", "PENDING"],
    "COMPLETED": [],
    "FAILED": ["PENDING", "RUNNING"],
    "SKIPPED": ["PENDING"],
}


async def create_task(db: AsyncSession, title: str, description: str | None = None, status: str = "Pending") -> Task:
    task = Task(title=title, description=description, status=status)
    db.add(task)
    await db.flush()
    db.add(ActivityLog(type="task.created", description=f"Task '{title}' created"))
    await db.commit()
    await db.refresh(task)
    return task


async def get_task(db: AsyncSession, task_id: str) -> Task | None:
    return await db.get(Task, task_id)


async def get_tasks(db: AsyncSession, status: str | None = None, page: int = 1, page_size: int = 20):
    q = select(Task)
    cq = select(func.count(Task.id))
    if status:
        q = q.where(Task.status == status)
        cq = cq.where(Task.status == status)
    q = q.order_by(Task.createdAt.desc()).offset((page - 1) * page_size).limit(page_size)
    result = await db.execute(q)
    tasks = list(result.scalars().all())
    total = (await db.execute(cq)).scalar_one()
    return tasks, total


async def update_task(db: AsyncSession, task_id: str, **kwargs) -> Task | None:
    task = await db.get(Task, task_id)
    if not task:
        return None
    for k, v in kwargs.items():
        if hasattr(task, k):
            setattr(task, k, v)
    if kwargs.get("status") == "Done":
        task.completedAt = datetime.utcnow()
    elif "status" in kwargs:
        task.completedAt = None
    db.add(ActivityLog(type="task.updated", description=f"Task '{task.title}' updated"))
    await db.commit()
    await db.refresh(task)
    return task


async def update_task_status(db: AsyncSession, task_id: str, new_status: str) -> Task | None:
    task = await db.get(Task, task_id)
    if not task:
        return None
    valid = VALID_STATUS_TRANSITIONS.get(task.status, [])
    if new_status not in valid:
        raise ValueError(f"Cannot transition from {task.status} to {new_status}. Valid: {valid}")
    task.status = new_status
    task.completedAt = datetime.utcnow() if new_status in ("Done", "COMPLETED") else None
    db.add(ActivityLog(type="task.updated", description=f"Task '{task.title}' status → {new_status}"))
    await db.commit()
    await db.refresh(task)
    return task


async def delete_task(db: AsyncSession, task_id: str) -> bool:
    task = await db.get(Task, task_id)
    if not task:
        return False
    await db.delete(task)
    db.add(ActivityLog(type="task.deleted", description=f"Task {task_id} deleted"))
    await db.commit()
    return True
